src/zon.zig

```diff
@@ -0,0 +1,661 @@
+const std = @import("std");
+const Zcu = @import("Zcu.zig");
+const Sema = @import("Sema.zig");
+const InternPool = @import("InternPool.zig");
+const Type = @import("Type.zig");
+const Zir = std.zig.Zir;
+const AstGen = std.zig.AstGen;
+const CompileError = Zcu.CompileError;
+const Ast = std.zig.Ast;
+const Allocator = std.mem.Allocator;
+const assert = std.debug.assert;
+const File = Zcu.File;
+const LazySrcLoc = Zcu.LazySrcLoc;
+const Ref = std.zig.Zir.Inst.Ref;
+const NullTerminatedString = InternPool.NullTerminatedString;
+const NumberLiteralError = std.zig.number_literal.Error;
+
+const LowerZon = @This();
+
+sema: *Sema,
+file: *File,
+file_index: Zcu.File.Index,
+
+/// Lowers the given file as ZON. `res_ty` is a hint that's used to add indirection as needed to
+/// match the result type, actual type checking is not done until assignment.
+pub fn lower(
+    sema: *Sema,
+    file: *File,
+    file_index: Zcu.File.Index,
+    res_ty: ?Type,
+) CompileError!InternPool.Index {
+    const lower_zon: LowerZon = .{
+        .sema = sema,
+        .file = file,
+        .file_index = file_index,
+    };
+    const tree = lower_zon.file.getTree(lower_zon.sema.gpa) catch unreachable; // Already validated
+    if (tree.errors.len != 0) {
+        return lower_zon.lowerAstErrors();
+    }
+
+    const data = tree.nodes.items(.data);
+    const root = data[0].lhs;
+    return lower_zon.expr(root, res_ty);
+}
+
+fn lazySrcLoc(self: LowerZon, loc: LazySrcLoc.Offset) !LazySrcLoc {
+    return .{
+        .base_node_inst = try self.sema.pt.zcu.intern_pool.trackZir(
+            self.sema.pt.zcu.gpa,
+            .main,
+            .{ .file = self.file_index, .inst = .main_struct_inst },
+        ),
+        .offset = loc,
+    };
+}
+
+fn fail(
+    self: LowerZon,
+    loc: LazySrcLoc.Offset,
+    comptime format: []const u8,
+    args: anytype,
+) (Allocator.Error || error{AnalysisFail}) {
+    @setCold(true);
+    const src_loc = try self.lazySrcLoc(loc);
+    const err_msg = try Zcu.ErrorMsg.create(self.sema.pt.zcu.gpa, src_loc, format, args);
+    try self.sema.pt.zcu.failed_files.putNoClobber(self.sema.pt.zcu.gpa, self.file, err_msg);
```

Comment:
I'm not sure whether these errors ought to go into `failed_files`. Once we change this to use type information properly (per a prior review comment), some of these errors are tied to the specific import site. Probably these should just forward through to `Sema.fail` instead, not touching the ZON file's state.